### ISSUE TITLE
Can the Attributes schema model be generalised to reflect division of semantic concerns thus richer knowledge representation?

### ISSUE THREAD

**Issue by RichardBruskiewich** [NCATSTranslator/ReasonerAPI#185]
The current ReasonerAPI (1.0) OpenAPI schema for the **Attribute** model in TRAPI is the following:

```yaml
    Attribute:
      type: object
      description: >-
        Generic attribute for a node or an edge that expands key-value pair
        concept by including a type of this attribute from a suitable ontology,
        a source of this attribute, and (optionally) a url with additional
        information about this attribute.
      properties:
        name:
          type: string
          description: >-
            Human-readable name or label for the attribute. If appropriate,
            should be the name of the semantic type term.
          example: PubMed Identifier
        value:
          example: 32529952
          description: >-
            Value of the attribute. May be any data type, including a list.
        type:
          $ref: '#/components/schemas/CURIE'
          description: >-
            CURIE of the semantic type of the attribute. For properties
            defined by the Biolink model this should be a biolink CURIE,
            otherwise, if possible, from the EDAM ontology. If a suitable
            identifier does not exist, enter a descriptive phrase here and
            submit the new type for consideration by the appropriate
            authority.
          example: EDAM:data_1187
        url:
          type: string
          description: >-
            Human-consumable URL to link out and provide additional information
            about the attribute (not the node or the edge).
          example: https://pubmed.ncbi.nlm.nih.gov/32529952
        source:
          type: string
          description: Source of the attribute, preferably as a CURIE prefix.
          example: UniProtKB
      required:
        - type
        - value
      additionalProperties: false
```

The limitation with this data model is that it doesn't adequately document context of interpretation for an attribute, when such context may be necessary and helpful. At best, it conflates such context with the nature of the attribute value.  This deficient is especially notable in attempts to represent so-called "evidence, provenance, confidence/context" information for TRAPI subgraphs being returned.

The key problem is that all of the fields, especially `type`, only refer to the value of the `attribute`. The `type` is expected to be a bioinformatics data type (i.e. semantics as typically qualified by the EDAM ontology). For example, a 'p value' is of data type `http://edamontology.org/data_1669`; however, the only interpretation of the model is that this statistic somehow a score related to the `Edge` but doesn't speak of its context of interpretation nor origin. It may be claimed that the statistic is somehow self descriptive but is that really true?

Here we will suggest modifications to the Attribute schema then we'll discuss the benefits.

1. Keep the existing field intact with respect to semantics relating to the `value` being reported but rename some fields to enforce this interpretation, as follows:

    1. Rename `type` to `value_type`
    1. Rename `name` to `value_type_name`
    1. Keep`value` as a string representation of the attribute value, whatever actual its data type
    1. Keep `url`: still refers to a resolvable URL to an identifier given as the `value` string
    1. Keep `source`: still refers to the immediate provenance (authority) which "owns" the `value`

1. Introduce two _additional_ fields targeting the _context of interpretation_ of the `value`: 

    1. Add `attribute_type` which should generally be a CURIE referencing the Biolink Model (usually, a Biolink `category` class), or perhaps a SEPIO term (for evidence, provenance and confidence metadata).
    1. Add `attribute_name` which should generally be the human readable name of the `attribute_type` CURIE entry

This is a first pass suggestion of Attribute model revision.  

A second level of elaboration may be helpful at the level of the `value` itself. The definition of `value` is open ended to allow "any datatype" including lists, which could included another JSON object or list of objects, depending on use case.

It might be helpful to formally document the possibility of a tree of attributes, e.g.

```yaml
    Attribute:
      type: object
...
      properties:
...
        value:
          description: >-
             Value of the attribute. May be any data type, including an list, including an
             embedded Attribute.  A list could be a array of Attributes.
          examples: 
             simple_value:
                 32529952
             simple_value_array:
                 - Hello
                 - World
             embedded_single_attribute_value:
                 attribute_type: SEPIO:0000130,
                 attribute_name: "assertionAuthoredBy",
                 value: "ORCID:12345",    
                 value_type: "wd:Q51044",
                 value_type_name: "ORCID ID"
                 url: https://orcid.org/12345
                 source: ORCID
             list_of_attribute_values:
                 - attribute_type: SEPIO:0000130,
                   attribute_name: "assertionAuthoredBy",
                   value: "ORCID:12345",    
                   value_type: "wd:Q51044",
                   value_type_name: "ORCID ID"
                   url: https://orcid.org/12345
                   source: ORCID
                 - attribute_type: SEPIO:0000135,
                   attribute_name: "dateAssertionAuthored",
                   value: "2018-05-12",    
                   value_type: "dc.date",
                   value_type_name: "date"
                   source: "MyFavoriteKnowledgeSource"
```

**Comment by vdancik** [comment]
Here is an example of attribute in the current format, can you specify how it would look under a new proposal?

```
{
    "attributes": [
        {
            "name": "mechanism_of_action",
            "source": "ChEMBL",
            "type": "NCIT:C54680",
            "value": "Cyclooxygenase inhibitor"
        }
    ]
}
```

**Comment by edeutsch** [comment]
I think Richard is correct that for many attributes it would be very useful to encode *what* is the attribute and *how* is it related.

So for Vlado's example, it could be this: (I renamed the key names that Richard proposed into what I thought seemed better, but I'm open to other possibilities)

```
{
    "attributes": [
        {
            "relation_id": "NCIT:C54680",
            "relation_name": "Mechanism of Action",
            "source": "ChEMBL",
            "type": "MESH:C0085387",
            "name": "Cyclooxygenase inhibitor",
            "url": "https://www.ncbi.nlm.nih.gov/medgen/43164"
        }
    ]
}
```

I assume that the scenario is that the KP is asserting that drug A treats disease B. But the KP wants to specify an attribute of that "treats" edge. The concept that we want as an attribute is "cyclooxygenase inhibitor", which happens to have an id MESH:C0085387. So now I have have a thing that is an attribute of the edge "treats" but we want to explain *how* it is related (its "relation"), and in this case it is NCIT:C54680 "Mechanism of Action".

I did not include "value" because that kind of thing (Cyclooxygenase inhibitor) doesn't take a value

Framing one of Richard's examples in this framework, I propose:
```
{
    "attributes": [
        {
            "relation_id": "SEPIO:0000130",
            "relation_name": "assertionAuthoredBy",
            "source": "ChEMBL",
            "type": "wd:Q51044",
            "name": "ORCID ID",
            "value": "ORCiD:0000-0001-8732-0928",
            "url": "https://orcid.org/0000-0001-8732-0928"
        }
    ]
}
```

Or maybe better would be:
```
{
    "attributes": [
        {
            "relation_id": "SEPIO:0000130",
            "relation_name": "assertionAuthoredBy",
            "source": "ChEMBL",
            "type": "ORCiD:0000-0001-8732-0928",
            "name": "Eric Deutsch",
            "url": "https://orcid.org/0000-0001-8732-0928"
        }
    ]
}
```

if we follow a distinction of an attribute that is itself an identifier that takes no value.

Perhaps an example of something with a value might be:

```
{
    "attributes": [
        {
            "relation_id": "SEPIO:0000187",
            "relation_name": "statement confidence level",
            "source": "ChEMBL",
            "type": "EDAM:data_1169",
            "name": "p-value",
            "value": "0.00343"
        }
    ]
}
```

The name value type is a p-value of 0.00343 like we had before, but the difference is that the relation_id and _name explains what the p-value really means in this context. (not a great example in this case)

In principle, this could allow you to do something like:

```
{
    "attributes": [
        {
            "relation_id": "SIO:000206",
            "relation_name": "is supported by",
            "source": "ChEMBL",
            "type": "EDAM:data_1187",
            "value": "PMID:2012345"
        },
        {
            "relation_id": "SIO:000207",
            "relation_name": "is disputed by",
            "source": "ChEMBL",
            "type": "EDAM:data_1187",
            "value": "PMID:2098765"
        }
    ]
}
```

These examples are not all fully consistent. Somehow we need to figure out the best way to encode an attribute that is a concept with a CURIE itself that takes no value (e.g. a PMID) vs. a concept that can be described with a CURIE but takes a value that cannot have a curie (like 0.00343). so this still needs some refinement.

What do you think?

**Comment by RichardBruskiewich** [comment]
Thanks @edeutsch (cc: @mbrush, @cmungall and @deepakunni3),

>So for Vlado's example, it could be this: (I renamed the key names that Richard proposed into what I thought seemed better, >but I'm open to other possibilities)
>
>{
>    "attributes": [
>        {
>            "relation_id": "NCIT:C54680",
>            "relation_name": "Mechanism of Action",
>            "source": "ChEMBL",
>            "type": "MESH:C0085387",
>            "name": "Cyclooxygenase inhibitor",
>            "url": "https://www.ncbi.nlm.nih.gov/medgen/43164"
>        }
>    ]
>}

I think I know what you are thinking: essentially, implying linkage of TRAPI Attributes back to the nomenclature of Translator knowledge graphs, essentially, the nomenclature of the Biolink Model.   

I'm ok with renaming the key names accordingly, but let's carefully ponder their meaning in the context of the Biolink Model.

The Biolink Model is largely composed of four metadata concepts: `type`, `mixin`, `slot` and `class`.

A Biolink `type` corresponds a data type primitive and I leave out the `mixin` for now since it is a more ephemeral creature, kind of like a `class` but more abstract.  I won't consider either of them further here although a Biolink `mixin` may sometimes play the role of an abstract class in the context of the model, and the Biolink `type` could be relevant sometimes for specifying the TRAPI Attribute `type` property of the `value`.

For now, we mainly focus on Biolink `class` and `slot` concepts here.

The two major root classes - `biolink:NamedThing` and `biolink:Association` - correspond to the Nodes and Edges of our knowledge graphs (hence correspond to TRAPI (Q)Node and (Q)Edge schema models, respectively).   In the Biolink Model, the `biolink:relation` is a slot of `biolink:Association` but so is `biolink:predicate`, thus these too are already directly represented in TRAPI (on in the (Q)Edge schemata).   

The Biolink Model also has a `biolink:Attribute` class which is now deliberately aligned to the TRAPI Attribute schema definition (as of PR https://github.com/biolink/biolink-model/pull/539). The `biolink:NamedThing` and `biolink:Association` classes may both also have lists of  `biolink:Attribute` hanging off of them, like instances of TRAPI (Q)Node and (Q)Edge data.

However, upon navigating the Biolink Model it can be noted that various subclass descendants of `biolink:NamedThing` and `biolink:Association` may have _other_ slots defined, specific to the use cases they embody.  We have also not yet (at least, not formally) specified Evidence Provenance Confidence metadata classes and slots in the Biolink Model, but EPC will likely eventually be represented with a suitable subset of Biolink classes and slots.

We obviously have the use case here of representing most `biolink:NamedThing` and `biolink:Association` slots - _other than `predicate` and `relation`_ - as TRAPI attributes as well.

Which leaves me to suggest that the names `slot_id` and `slot_name` rather than `relation_id` and `relation_name` be used in the TRAPI Attribute model since using `relation_id` and `relation_name` in the Attribute would be conflate the `relation` `biolink:Association` slot (a.k.a. `(Q)Edge` property) by the same name.  

On the other hand, using `slot_id` and `slot_name` for these Attribute property names would make an solid unambiguous conceptual bridge of intent and usage back to the Biolink Model.

The only additional point I would make here is that, perhaps, in addition to EDAM ontology defined attribute `type` values, we may wish to allow (as use cases dictate) any child subclass descendant of `biolink:NamedThing` as well, since in several cases, the value of a slot in `biolink:NamedThing` or `biolink:Association`, may sometimes be such an instance of a knowledge graph node (of a specific `biolink:category`). Thus, the `type` of the `value` of such (TRAPI Attribute expressed) slots could be a `biolink:category` as well.  This may prove especially true in the future, once more EPC metadata is explicitly modelled inside the Biolink Model.

**Comment by RichardBruskiewich** [comment]
OK, @edeutsch (again, cc: @cmungall, @deepakunni3 and @mbrush), despite the verbosity of my previous comment, it actually only addressed one facet of your feedback: your proposed TRAPI Attribute schema property names `relation_id` and `relation_name`. 

Now, I need to respond in more depth to other aspects of your feedback.

>I assume that the scenario is that the KP is asserting that drug A treats disease B. But the KP wants to specify an attribute of >that "treats" edge. The concept that we want as an attribute is "cyclooxygenase inhibitor", which happens to have an id >MESH:C0085387. So now I have have a thing that is an attribute of the edge "treats" but we want to explain how it is related (its >"relation"), and in this case it is "Cyclooxygenase inhibitor".   

Vlado's example is, _a priori_,  a bit tricky, especially in the context of your reply. I'm going to brainstorm various ways of representing the 'fact' I think is being expressed by the example, assuming, as you indicate, that the basic Edge assertion is that **drug A treats disease B**.

## Option A

  Given that you renamed the attribute fields to `relation_id` and `relation_name`, let us us ask: why not set the (Q)Edge `relation` property itself to `MESH:C0085387`?  One doesn't actually need the `relation_name` here since one can look up `MESH:C0085387`  find the label "Cyclooxygenase inhibitor". However, "_what about **Mechanism of Action**?_" you say. Isn't "Mechanism of Action" a matter of default implication for treats ["Cyclooxygenase inhibitor"]. 

There's actually a small "slight of hand" here which, ironically, Matt brought up earlier today in another Biolink discussion: should an attribute of, say, a subject, be injected into the 'relationship' Edge?  Otherwise put, **A treats [_is a_ "Cyclooxygenase inhibitor" _of_] B** or perhaps, **B biolink:treated_by [_has_ "Cyclooxygenase inhibitor"] A**.

```
Edge.subject: A
Edge.predicate: treats
Edge.relation: MESH:C0085387  # "Cyclooxygenase inhibitor"
Edge.object: B
```

## Option B

Another option is to make the "Mechanism of Action" the `relation` and the "Cyclooxygenase inhibitor" an Edge.qualifier, as follows:

```
Edge.subject: A
Edge.predicate: treats
Edge.relation: NCIT:C54680   # "Mechanism of Action"
Edge.object: B
Edge.Attribute.slot_id: "biolink:qualifier"
Edge.Attribute.type: MESH:C0085387  # "Cyclooxygenase inhibitor"
```

As you may know, `biolink:qualifier` is a first class property of the `biolink:Association` class but TRAPI effectively pushes it down to an Attribute of a (Q)Edge. That's fine, if we use the property names `slot_id` and `slot_name` in Attribute.

If you think about this a minute, maybe we don't (again) need `slot_name` since that can be looked up in the Biolink Model itself.  Alternately (the @deepakunni3 et al. may wish to comment) perhaps if we only have one property in the Attribute to hold a Biolink CURIE, it could just be `slot` (where the definition stipulates "Biolink Model entity CURIE"). And, of course, this property would be optional, if in fact, the given Attribute has no direct mapping onto the Biolink Model, or it may conceivably be permitted to hold a non-Biolink CURIE in this context, if sensible (e.g. a SEPIO term CURIE).

**Comment by RichardBruskiewich** [comment]
@edeutsch, next example:

```
{
    "attributes": [
        {
            "relation_id": "SEPIO:0000130",
            "relation_name": "assertionAuthoredBy",
            "source": "ChEMBL",
            "type": "ORCID:0000-0001-8732-0928",
            "name": "Eric Deutsch",
            "url": "https://orcid.org/0000-0001-8732-0928"
        }
    ]
}
```

I struggling with the notion that `type` should be set to an instance of ORCID ID. It's not really the `type` of value. I'd much rather see the `EDAM_DATA:1881` - "Author ID" or similar as the `type` and keep the `value` set to `ORCID:0000-0001-8732-0928`.  Better yet, ORCID should be offered to EDAM as a new term (just submitted the issue, LOL) since ORCID identifiers are helpful (meta)data in the biomedical community perhaps?

The `url` is fine. The `name` is convenient but superfluous (one can look it up later, using the `value`).

**Comment by RichardBruskiewich** [comment]
@edeutsch, your other two examples look pretty close to the mark (except, of course, for the initial naming issue about which I've commented). I would like to leave @mbrush do the honours of commenting further on those examples.

**Comment by mbrush** [comment]
Hi all.  Lots to address here (some of it a bit off the main topic) - but will try to keep as concise as possible, and push some tangents to new tickets.
 
1. Regarding the assertion being made in the Vlado's association, he shared the full example with me earlier and it says that a **Chemical (aspirin)** - *affects* -> **Gene (COX2)**.  Nothing to do with a disease. 
 
2. Regarding the idea of 'Mechanism of Action' terms living in the *relation* slot - this is a bit off the main topic of this thread so I will take the discussion to a new ticket [here](https://github.com/biolink/biolink-model/issues/570). It discusses how we might represent Mechanism of Action in the context of a Chemical-treats-Disease and a Chemical-affects-Gene association.  
 
3. To me the question of whether we want to pass along human readable names of curies in a TRAPI message (e.g. that the label of EDAM:data_1169 class is "p-value") comes down to whether we want humans to be able to understand the content of the message - without having to do lookups form external sources.  If this is important there are several structures we could use to do this.  If not, and we leave grabbing human-readable names to a downstream users / presentation layer of the architecture, it simplifies the message to leave them out. Curious what people think.

**Comment by mbrush** [comment]
Regarding the naming of the Attribute *relation_id* / *slot_id* field, as debated in the earlier comments above,  ultimately what I care about more than the name is that its meaning/intent is clear.  But a name based on a principled understanding of the Attribute object would go a long way here. 
 
To my way of thinking (and I believe Chris Mungall's based on recent convos), an 'Attribute' object hanging from an Association/Edge is essentially a **reified association slot/edge property**. This reification lets us push the semantics of the property into the data, so that the TRAPI schema itself doesn’t need to change to accommodate new types of information on an edge.  So instead of a named *Edge.authoredBy* property, we create an Attribute object that represents/reifies this slot and its value, and we define a slot in the Attribute object to types the type of named attribute it reifies (in this case "authoredBy").
 
If we choose to name this reified object "Attribute", as TRAPI has done, I think it makes sense to name the Attribute slot that describes the edge property that the Attribute reifies as the '*attribute_type*'.   If, however, we named the reified object 'Property' instead of 'Attribute', we could call this slot '*property_type*'. And if we named it 'Slot', we could call it '*slot_type*'.  Point is, I feel like a **connection** between the name we give the reified object and the name of field that defines what the object reifies is helpful, given a reification paradigm. 

Just my 2c - but again, I don’t have strong feelings here as long as things are consistent and clear.   W.r.t. other names proposed for this field, I see Richard's  point that naming it '*relation_id*' could be confusing given that Edges already hold a named '*relation*' field.  Naming it '*slot_id*', while preferrable, could still be confusing given that we call the reified object an 'Attribute', and not a 'Slot'.  If we did this our documentation would have to be clear that this name is based on the understanding that the Attribute object represents a reified slot, and hence the '*slot_id*' field describes the slot that is being reified. 
 
Finally, FWIW we could adopt/extend the[ FHIR 'Extension' ](https://www.hl7.org/fhir/extensibility.html) element and naming conventions - as this is an established and successful approach to address our use case.  I think it does the same thing we are trying to do here.
